### ISSUE TITLE
Fix roundness in rects

### ISSUE THREAD

**Issue by rlafuente** [shoebot/shoebot#361]
The `roundness` parameter works just like Nodebox, but values above 0.5 give glitchy results. We can find a better way to deal with this. Pasted from the #shoebot chat:

![Screenshot_2021-03-25_22-07-53](https://user-images.githubusercontent.com/159868/112550925-07d79580-8db8-11eb-82d9-fd0dfd89a201.png)

so what we have comes straight from nodebox, and the picture makes it somewhat clear why things go wrong with roundness > 0.5
their docs don't mention expected values for roundness, so i think they just didn't care much for it
i'd like at some point to review this behaviour because i'm sure there's ways to round things beyond 0.5 without weirdness
here what happens is that the "lineto" portion of the rect (between the two red dots) get smaller and smaller as you approach 0.5
after 0.5 it becomes negative and that weird twisting happens; we'd just need a check so that, beyond roundness=0.5, only the control points move further, not the path/destination points
finally -- drawbot doesn't have the roundness parameter in rects, so this new behaviour would be on us
(i'm not too worried about breaking nodebox compatibility here, since the idea is to keep the same behavior for roundness < 0.5, and implement a better way for values above 0.5, which no one should be using anyway since it's clearly glitchy)

**Comment by rlafuente** [comment]
Code for the test script that generated the above output:

```python
size(600,300)

def circlepoint(x,y,sz=4):
    circle(x-sz/2,y-sz/2, sz,sz)

def rect_marks(x,y,w,h,roundness=0):
    # rectangle with marks on the first lineto extremes
    rect(x,y,w,h,roundness)
    curve = min(w * roundness, h * roundness)
    fill("#dd1111", 0.5)
    nostroke()
    circlepoint(x+curve, y)
    circlepoint(x+w-curve, y)

translate(50,50)
font("Inter", 12)

counter = 0
for x,y in grid(5,2,100,100):
    fill("#2211bb", 0.5)
    stroke(0.5)
    roundness = counter / 10
    if roundness <= 1:
        rect_marks(x, y, 80, 80, roundness)
    fill(0.5)
    text(str(roundness), x, y+95)
    counter += 1
```

**Comment by rlafuente** [comment]
Also of interest is the way Pippin made a round rectangle in this [Cairo example](https://cairographics.org/samples/curve_rectangle/)

**Comment by stuaxo** [comment]
Maybe it helps to think about what roundness of 1.0 means, should you be able end up with a circle under some circumstances, for instance ?

**Comment by rlafuente** [comment]
yeah, we could go multiple ways. I'll keep this one in the backburner and look at this again after 1.4.

**Comment by rlafuente** [comment]
Perhaps adding a note in the docs about how it should be used between 0 and 0.5?

**Comment by stuaxo** [comment]
Either a note, or maybe add some code to clamp it to 0.0 - 0.5 ?

**Comment by karstenw** [comment]
> Either a note, or maybe add some code to clamp it to 0.0 - 0.5 ?

IMHO neither.

The correct way would be to rewrite it. The function should give a circle or oval for roundness==1 and guarding against the range 0.0-1.0.

Using values from -0.5 to 1.5 gives interesting results.
![rectroundness](https://user-images.githubusercontent.com/652583/189318448-a6e5ebf3-ac70-4203-aa2f-7efef5ed250e.png)

**Comment by rlafuente** [comment]
This is a really useful diagram, thank you! Looking at this, the "reasonable" range would indeed be 0-0.5, after which the handles start doing ugly things.

I think the simplest fix would be to accept values from 0 to 1, which are mapped to the 0-0.5 range. You can still provide strange values like -1 or 2 and get strange results.

If there's no objection, I can go update the code and add a note to the docs about the new behaviour

**Comment by stuaxo** [comment]
That sounds good to me.

0 - 1.0 makes sense, with the setting being called "roundness".

I started to wonder if we want to give the user more control, but things get complex then and I wonder if we start re-implementing something a little like CSS styling.
